src/backend/api/meals.js

```diff
@@ -1,15 +1,199 @@
+    }
+    // Only specific number of meals
+    if (limit) {
+      const mealLimit = parseInt(limit)
+      meals = await knex('meal').select('*').limit(mealLimit)
+    }
+    // Get meals that still has available reservations
+    if (availableReservations) {
+      meals = await knex('meal')
+        .leftJoin('reservation', 'meal.id', '=', 'reservation.meal_id')
+        .select(
+          'meal.id',
+          'meal.title',
+          'meal.description',
+          'meal.location',
+          'meal.when',
+          'meal.created_date',
+          'meal.price',
+          'meal.max_reservations',
+          knex.raw(
+            'COALESCE(SUM(reservation.number_of_guests),0) as total_guests'
+          ),
+          knex.raw(
+            '(meal.max_reservations-COALESCE(SUM(reservation.number_of_guests),0)) AS available_reservation'
+          )
+        )
+        .groupBy('meal.id')
+        .having(
+          knex.raw(
+            '(max_reservations > COALESCE(SUM(reservation.number_of_guests),0))'
+          )
+        )
+        .groupBy('meal.id')
+    }
+    response.json(meals)
+  } catch (error) {
+    throw error
+  }
+})
+
+//GET. Returns meal by id
+router.get('/:id', async (request, response) => {
+  try {
+    const mealId = Number(request.params.id)
+    if (isNaN(mealId)) {
+      response.status(400).json({
+        error: "Meal wasn't found.",
+        hint: 'Id should be a number',
+      })
+      return
+    } else {
+      const meals = await knex('meal')
+      const isFoundId = meals.some((meal) => meal.id === mealId)
+      if (!isFoundId) {
+        response.status(400).json({
+          error: "Meal wasn't found.",
+          hint: 'Meal with this id does not exist',
+        })
+        return
+      }
+      const mealById = await knex('meal').select('*').where({ id: mealId })
+      response.json(mealById)
+    }
+  } catch (error) {
+    throw error
+  }
+})
+
+//POST.	Adds a new meal
+router.post('/', async (request, response) => {
+  try {
+    const context = request.body
+    if (Object.keys(context).length === 0 || context === {}) {
+      response.statusCode(422).json({ message: 'Bad input' })
+      return
+    }
+    if (context.title === '') {
+      response.json({ message: 'Give me title' })
+      return
+    }
+    if (context.description === '') {
+      response.json({ message: 'Give me description' })
+      return
+    }
+    if (context.location === '') {
+      response.json({ message: 'Give me location' })
+      return
+    }
+    if (context.when === '') {
+      response.json({ message: 'Give me date' })
+      return
+    }
+    if (context.max_reservations === '') {
+      response.json({ message: 'Give number of guests' })
+      return
+    }
+    if (context.price === '') {
+      response.json({ message: 'Give me max price' })
+      return
+    }
+    if (context.created_date === '') {
+      response.json({ message: 'Give me created date' })
+      return
+    }
```

Comment:
Very good again that you're thinking about error handling and validation! But what happens if your API will accept 100, or 1000 parameters (I'm not saying it's a great idea!)

Instead of having so many nested if clauses, you could have a single one, and have a way shorter and readable code.

```js
const requiredParameters = ['title', 'description', 'location', 'max_reservations', 'price', 'created_data'];

requiredParameters.map( param => {
  if (context[param] === '') {
    response.statusCode(422).json({ message: `Bad input; Required param: $param` })
  }
} );


This way, we have an array of required parameters, and we iterate over them, checking if any is empty. If you want at some point to add an extra parameter, you can just add it to the list and you're done 👌🏼

Reply:
I have changed  a little 🤏. Because of this. SyntaxError: await is only valid in async functions and the top level bodies of modules